fog/ledger/server/tests/router_connection.rs

```diff
@@ -0,0 +1,910 @@
+// Copyright (c) 2018-2022 The MobileCoin Foundation
+
+//! Integration tests at the level of the fog ledger connection / fog ledger
+//! grpc API
+
+use futures::executor::block_on;
+use grpcio::ChannelBuilder;
+use mc_account_keys::{AccountKey, PublicAddress};
+use mc_api::watcher::TimestampResultCode;
+use mc_attest_net::{Client as AttestClient, RaClient};
+use mc_attest_verifier::{MrSignerVerifier, Verifier, DEBUG_ENCLAVE};
+use mc_blockchain_types::{BlockSignature, BlockVersion};
+use mc_common::{
+    logger::{test_with_logger, Logger},
+    time::SystemTimeProvider,
+    ResponderId,
+};
+use mc_crypto_keys::{CompressedRistrettoPublic, Ed25519Pair};
+use mc_fog_api::{ledger::TxOutResultCode, ledger_grpc::KeyImageStoreApiClient};
+use mc_fog_ledger_connection::{
+    Error, FogMerkleProofGrpcClient, FogUntrustedLedgerGrpcClient, KeyImageResultExtension,
+    LedgerGrpcClient, OutputResultExtension,
+};
+use mc_fog_ledger_enclave::LedgerSgxEnclave;
+use mc_fog_ledger_server::{
+    sharding_strategy::EpochShardingStrategy, KeyImageStoreServer, LedgerRouterConfig,
+    LedgerRouterServer, LedgerStoreConfig, ShardingStrategy,
+};
+use mc_fog_test_infra::get_enclave_path;
+use mc_fog_uri::{ConnectionUri, FogLedgerUri, KeyImageStoreUri};
+use mc_ledger_db::{test_utils::recreate_ledger_db, Ledger, LedgerDB};
+use mc_transaction_core::{
+    membership_proofs::compute_implied_merkle_root, ring_signature::KeyImage, tokens::Mob, Amount,
+    Token,
+};
+use mc_util_from_random::FromRandom;
+use mc_util_grpc::{ConnectionUriGrpcioChannel, GrpcRetryConfig, CHAIN_ID_MISMATCH_ERR_MSG};
+use mc_util_test_helper::{CryptoRng, RngCore, RngType, SeedableRng};
+use mc_util_uri::AdminUri;
+use mc_watcher::watcher_db::WatcherDB;
+use portpicker::pick_unused_port;
+use std::{
+    collections::HashMap,
+    path::PathBuf,
+    str::FromStr,
+    sync::{Arc, RwLock},
+    thread::sleep,
+    time::Duration,
+};
+use tempdir::TempDir;
+use url::Url;
+
+const TEST_URL: &str = "http://www.my_url1.com";
+
+const OMAP_CAPACITY: u64 = 128 * 128;
+
+const GRPC_RETRY_CONFIG: GrpcRetryConfig = GrpcRetryConfig {
+    grpc_retry_count: 3,
+    grpc_retry_millis: 20,
+};
+
+fn setup_watcher_db(logger: Logger) -> (WatcherDB, PathBuf) {
+    let url = Url::parse(TEST_URL).unwrap();
+
+    let db_tmp = TempDir::new("wallet_db").expect("Could not make tempdir for wallet db");
+    WatcherDB::create(db_tmp.path()).unwrap();
+    let watcher = WatcherDB::open_rw(db_tmp.path(), &[url], logger).unwrap();
+    let watcher_dir = db_tmp.path().to_path_buf();
+    (watcher, watcher_dir)
+}
+
+// Test that a fog ledger connection is able to get valid merkle proofs by
+// hitting a fog ledger server
+#[test_with_logger]
+fn fog_ledger_merkle_proofs_test(logger: Logger) {
```

Comment:
Can we remove the configuration or move some of the set up logic to helper functions / a helper module? Putting all this configuration in the test makes it harder to read and leads to  duplication in tests below.

Reply:
I didn't write these tests originally, they're just copies of `connection.rs`. You're absolutely right about the structure of the tests not being great. I'll file an issue for fixing it but I don't want to block this PR on it.

Reply:
#3168